File: 2856-count-complete-subarrays-in-an-array/count-complete-subarrays-in-an-array.cpp
class Solution {
public:
    int countCompleteSubarrays(vector<int>& nums) 
    {
        unordered_set<int> set(nums.begin(), nums.end());

        unordered_map<int, int> map;
        int i = 0, count = 0, n = nums.size();
        for(int j = 0; j < n; j++)    
        {
            map[nums[j]]++;
            while(map.size() == set.size())
            {
                count += n - j;

                map[nums[i]]--;
                if(map[nums[i]] == 0)
                map.erase(nums[i]);

                i++;
            }
        }
        return count;
    }
};

File: 2856-count-complete-subarrays-in-an-array/count-complete-subarrays-in-an-array.java
class Solution {
    public int countCompleteSubarrays(int[] nums) {
        // Set to count distinct elements in the entire array
        Set<Integer> unique = new HashSet<>();
        for (int num : nums) {
            unique.add(num);
        }

        int n = nums.length;
        int count = 0;
        Map<Integer, Integer> freqMap = new HashMap<>();
        int i = 0;

        for (int j = 0; j < n; j++) {
            freqMap.put(nums[j], freqMap.getOrDefault(nums[j], 0) + 1);

            while (freqMap.size() == unique.size()) {
                // all subarrays starting at i and ending at j...n-1 are valid
                count += n - j;

                freqMap.put(nums[i], freqMap.get(nums[i]) - 1);
                if (freqMap.get(nums[i]) == 0) {
                    freqMap.remove(nums[i]);
                }
                i++;
            }
        }

        return count;
    }
}
